### PR TITLE
[incubator/ibm-blockchain-chaincode] Add IBM Blockchain Chaincode

### PR DESCRIPTION
**Pull request by some-ibmer**
Source code located in the following repository:

- https://github.com/IBM-Blockchain/ibm-container-service

### CODE REVIEW

File: incubator/ibm-blockchain-chaincode/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: ibm-blockchain-chaincode
+version: 1.0.0
+description: IBM Blockchain Chaincode Chart
+keywords:
+  - blockchain
+home: https://ibm-blockchain.github.io
+sources:
+  - https://github.ibm.com/IBM-Blockchain/ibm-container-service
+maintainers:
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/Chart.yaml]: Use Github usernames.

File: incubator/ibm-blockchain-chaincode/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+name: ibm-blockchain-chaincode
+version: 1.0.0
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/Chart.yaml:2]: Add `appVersion`.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR and sorry for the delay.

Please follow our best practices and make sure you add standard labels on all resources.

https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/labels.md

Note that we've started namespacing templates (see #1785). Please apply to your PR.

File: incubator/ibm-blockchain-chaincode/Chart.yaml (hunk 2 above)
Comment (scottgigawatt) [reply incubator/ibm-blockchain-chaincode/Chart.yaml:2]: Done.

File: incubator/ibm-blockchain-chaincode/Chart.yaml (hunk 1 above)
Comment (scottgigawatt) [reply incubator/ibm-blockchain-chaincode/Chart.yaml]: Done.

**Comment by some-ibmer** [comment]
@unguiculus, changes have been added.

**Comment by unguiculus** [comment]
Please consider my comment regarding namespacing and best practices. Once you have your templates namespaced, follow this example for naming resources and labels:

```yaml
name: {{ template "ibm-blockchain-chaincode.fullname" . }}
labels:
  app: {{ template "ibm-blockchain-chaincode.name" . }}
  chart: {{ .Chart.Name }}-{{ .Chart.Version }}
  release: {{ .Release.Name }}
  heritage: {{ .Release.Service }}
```

**Comment by some-ibmer** [comment]
@unguiculus, I've added the namespaces.

**Comment by unguiculus** [comment]
Let me point out our best practices again. And please read our review guidelines which give you a good overview across what we expect.

https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

**Comment by some-ibmer** [comment]
@unguiculus, the templates should be namespaced now.

File: incubator/ibm-blockchain-chaincode/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+$ helm install stable/ibm-blockchain-chaincode
+```
+
+## Introduction
+
+This chart bootstraps a [ibm-blockchain](https://ibm-blockchain.github.io) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
+
+## Prerequisites
+
+- Kubernetes 1.4+ with Beta APIs enabled
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/README.md]: We officially support the current and previous minor versions, i. e. 1.7+

File: incubator/ibm-blockchain-chaincode/templates/_helpers.tpl [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "name" -}}
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/templates/_helpers.tpl]: Templates not namespaced, but namespaced templates used in resources.

File: incubator/ibm-blockchain-chaincode/templates/chaincode_install_org2peer1.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  volumes:
+  - name: shared
+    persistentVolumeClaim:
+     claimName: shared
+
+  containers:
+  - name: chaincodeinstall-org2peer1
+    image: ibmblockchain/fabric-tools:1.0.0
+    imagePullPolicy: {{.Values.blockchain.pullPolicy}}
+    command: ["sh", "-c", "while [ ! -f /shared/status_joinchannel_org2peer1_complete ]; do echo Waiting for joinchannel; sleep 2; done; git clone -b v1.0.0 https://github.com/hyperledger/fabric $GOPATH/src/github.com/hyperledger/fabric/ && peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02/ && touch /shared/status_chaincodeinstall_complete"]
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/templates/chaincode_install_org2peer1.yaml]: How about wrapping the long line?

```yaml
command:
  - sh
  - -c
  - |
    while [ ! -f /shared/status_joinchannel_org2peer1_complete ]; do 
      echo Waiting for joinchannel
      sleep 2
    done

    git clone -b v1.0.0 https://github.com/hyperledger/fabric $GOPATH/src/github.com/hyperledger/fabric/
    peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02/
    touch /shared/status_chaincodeinstall_complete
```

File: incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  restartPolicy: "Never"
+  volumes:
+  - name: shared
+    persistentVolumeClaim:
+     claimName: shared
+
+  containers:
+  - name: chaincodeinstantiate
+    image: ibmblockchain/fabric-tools:1.0.0
+    imagePullPolicy: {{.Values.blockchain.pullPolicy}}
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml]: Make formatting consistent:
`{{.Values.blockchain.pullPolicy}}` -> `{{ .Values.blockchain.pullPolicy }}`

File: incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+spec:
+  restartPolicy: "Never"
+  volumes:
+  - name: shared
+    persistentVolumeClaim:
+     claimName: shared
+
+  containers:
+  - name: chaincodeinstantiate
+    image: ibmblockchain/fabric-tools:1.0.0
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml]: Image should be configurable

File: incubator/ibm-blockchain-chaincode/values.yaml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  # lines, adjust them as necessary, and remove the curly braces after 'resources:'.
+  # limits:
+  #  cpu: 100m
+  #  memory: 128Mi
+  # requests:
+  #  cpu: 100m
+  #  memory: 128Mi
+
+blockchain: 
+  pullPolicy: Always
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/values.yaml]: Default should be `IfNotPresent`

File: incubator/ibm-blockchain-chaincode/templates/NOTES.txt [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+{{- range .Values.ingress.hosts }}
+  http://{{ . }}
+{{- end }}
+{{- else if contains "NodePort" .Values.service.type }}
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "fullname" . }})
+  export NODE_IP=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath="{.items[0].status.addresses[0].address}")
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get svc -w {{ template "fullname" . }}'
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/templates/NOTES.txt]: Templates are not namespaced.

File: incubator/ibm-blockchain-chaincode/README.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+```console
+$ helm delete my-release
+```
+
+The command removes all the Kubernetes components associated with the chart and deletes the release.
+
+## Configuration
+
+The following tables lists the configurable parameters of the IBM Blockchain chart and their default values.
+
```
Comment (unguiculus) [inline incubator/ibm-blockchain-chaincode/README.md]: Table incomplete.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/ibm-blockchain-chaincode/values.yaml (hunk 8 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/values.yaml]: Fixed

File: incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml (hunk 6 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml]: fixed

File: incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml (hunk 7 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/templates/chaincode_instantiate.yaml]: fixed

File: incubator/ibm-blockchain-chaincode/templates/chaincode_install_org2peer1.yaml (hunk 5 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/templates/chaincode_install_org2peer1.yaml]: fixed

File: incubator/ibm-blockchain-chaincode/README.md (hunk 3 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/README.md]: fixed

File: incubator/ibm-blockchain-chaincode/README.md (hunk 10 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/README.md]: Added additional values to table.  Hopefully, the table is now complete?

File: incubator/ibm-blockchain-chaincode/templates/_helpers.tpl (hunk 4 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/templates/_helpers.tpl]: I think this is fixed now?

File: incubator/ibm-blockchain-chaincode/templates/NOTES.txt (hunk 9 above)
Comment (some-ibmer) [reply incubator/ibm-blockchain-chaincode/templates/NOTES.txt]: I think this is fixed now?

**Comment by some-ibmer** [comment]
@unguiculus, this PR should be up-to-date with the requested changes.  Please let me know if additional changes are required.

**Comment by some-ibmer** [comment]
@unguiculus, I'm not sure why the ci/circleci cannot find the tests [here](https://github.com/kubernetes/charts/pull/2071/files#diff-8b0b63bd82336af5bb8e9fb8f1de2cfa).

**Pull request closed without merging**